cpp/dolfin/io/XDMFFile.cpp

```diff
@@ -1596,9 +1720,33 @@ XDMFFile::read_checkpoint(std::shared_ptr<const function::FunctionSpace> V,
   const std::array<std::int64_t, 2> input_vector_range
       = MPI::local_range(_mpi_comm.comm(), num_global_dofs);
 
+#ifdef PETSC_USE_COMPLEX
+  // Read real component of function vector
+  std::vector<double> real_vector = get_dataset<double>(
+      _mpi_comm.comm(), vector_dataitem, parent_path, input_vector_range);
+
+  // Get extra FE attribute of the imaginary component
+  pugi::xml_node imag_vector_dataitem
+      = imag_fe_attribute_node.select_node("DataItem[position()=2]").node();
+  assert(imag_vector_dataitem);
+
+  // Read imaginary component of function vector
+  std::vector<double> imag_vector = get_dataset<double>(
+      _mpi_comm.comm(), imag_vector_dataitem, parent_path, input_vector_range);
+
+  assert(real_vector.size() == imag_vector.size());
+
+  // Compose complex function vector
+  std::vector<PetscScalar> vector;
+  vector.reserve(real_vector.size());
+  std::transform(begin(real_vector), end(real_vector), begin(imag_vector),
+                 std::back_inserter(vector),
+                 [](double r, double i) { return r + i * PETSC_i; });
```

Comment:
The memory footprint of this is rather large. It seems we can again, as in the Hyperslab issue above, tell HDF5 to write the data directly into strided memory. See https://support.hdfgroup.org/HDF5/Tutor/selectsimple.html#def. Can be done in the next PR.